src/main/java/frc/robot/gamepieces/AbstractLayers/ShooterAL.java

```diff
@@ -0,0 +1,375 @@
+        flywheelFollower = null;
+
+        if (RobotMap.SHOOTER_FOLLOWER) {
+          LOGGER.info("Creating first set of follower motors");
+          flywheelFollower = new WPI_TalonSRX(RobotMap.SHOOTER_MOTOR_FOLLOWER_CHANNEL);
+          flywheelFollower.setNeutralMode(NeutralMode.Coast);
+        }
+
+        flywheel = new TalonSpeedControllerGroup("Shooter", ControlMode.Velocity, RobotMap.SHOOTER_SENSOR_INVERTED,
+            RobotMap.SHOOTER_MOTOR_INVERTED, flywheelLeader, flywheelFollower);
+
+        flywheel.pidf(RobotMap.SHOOTER_PID_SLOT_DRIVE, RobotMap.SHOOTER_P, RobotMap.SHOOTER_I, RobotMap.SHOOTER_D,
+            RobotMap.SHOOTER_F, RobotMap.VELOCITY_MULTIPLIER_SHOOTER);
+      } else {
+        flywheel = new TalonSpeedControllerGroup();
+      }
+
+      if (RobotMap.HAS_SHOOTER_TRIGGER) {
+        triggerMotor = new WPI_TalonSRX(RobotMap.TRIGGER_MOTOR_CHANNEL);
+        trigger = new TalonSpeedControllerGroup("Trigger", ControlMode.PercentOutput, false, RobotMap.TRIGGER_MOTOR_INVERTED, triggerMotor);
+      } else {
+        trigger = new TalonSpeedControllerGroup();
+      }
+
+      if (RobotMap.HAS_SHOOTER_HOOD) {
+        hoodLeader = new Servo(RobotMap.HOOD_PWM_PORT);
+        hoodFollower = null;
+
+        if (RobotMap.HOOD_FOLLOWER) {
+          hoodFollower = new Servo(RobotMap.HOOD_FOLLOWER_PWM_PORT);
+        }
+
+      } else {
+        hoodLeader = null;
+        hoodFollower = null;
+      }
+
+      if (RobotMap.HAS_SHOOTER_LEDS) {
+        leds = new AddressableLED(RobotMap.SHOOTER_LED_CHANNEL);
+        ledBuffer = new AddressableLEDBuffer(RobotMap.SHOOTER_LED_AMOUNT * (RobotMap.SHOOTER_DOUBLESIDE_LED? 2: 1));
+        leds.setLength(ledBuffer.getLength());
+
+        for (var i = 0; i < ledBuffer.getLength(); i++) {
+          ledBuffer.setRGB(i, 0, 0, 0);
+        }
+
+        leds.setData(ledBuffer);
+        leds.start();
+      } else {
+        leds = null;
+        ledBuffer = null;
+      }
+
+      instance = new ShooterAL(flywheel);
+      instance.stop();
+    }
+    return instance;
+  }
+
+
+  private ShooterAL(TalonSpeedControllerGroup flywheel) {
+    super("Telemetry", "Shooter");
+  }
+
+  public TalonSpeedControllerGroup getMotor() {
+    return flywheel;
+  }
+
+  public void stop() {
+    //LOGGER.error("Stop called");
+    flywheel.stopMotor();
+  }
+
+  public void setSpeed(double speed) {
+    this.speed = speed;
+    
+  }
+
+  public void rampToSpeed(double speed) {
+    if (flywheel != null && RobotMap.HAS_SHOOTER) {
+      double output = Math.max(-1.0, Math.min(1.0, speed));
+      flywheel.set(ControlMode.Velocity, output);
+      LOGGER.debug("the speed is {}", output);
+    }
+  }
+
+  public boolean atSpeed() {
+    double current = flywheel.velocity();
+    double target = flywheel.closedLoopTarget();
+    double error = current / target;
+
+    if (error <= 1 + RobotMap.SHOOTER_SPEED_TOLERANCE || error >= 1 - RobotMap.SHOOTER_SPEED_TOLERANCE) {
```

Comment:
This may be simpler:
`if (Math.abs(error - 1) > RobotMap.SHOOTER_SPEED_TOLERENCE) {`